lib/fca/fca.nit

```diff
@@ -0,0 +1,273 @@
+
+		return concepts
+	end
+end
+
+# Formal Concept
+#
+# A pair *(A,B)* is a formal concept of a FormalContext *(G, M, I)* provided that:
+#
+# * *A ⊆ G*,
+# * *B ⊆ M*,
+# * *A′ = B*, and
+# * *B′ = A*.
+#
+# Equivalently and more intuitively, *(A,B)* is a formal concept precisely when:
+#
+# * every object in *A* has every attribute in *B*,
+# * for every object in *G* that is not in *A*, there is some attribute in *B* that
+#   the object does not have,
+# * for every attribute in *M* that is not in *B*, there is some object in *A*
+#   that does not have that attribute.
+class FormalConcept[O: Object, A: Object]
+
+	# Concept attributes
+	var attributes = new HashSet[A]
+
+	# Concept objects
+	var objects = new HashSet[O]
+
+	# Is `self` a subconcept of `super_concept`?
+	#
+	# A concept C1 is a subconcept of C2 if C2 has all the objects of C1.
+	fun is_subconcept(super_concept: FormalConcept[O, A]): Bool do
+		if self == super_concept then return false
+		if objects.length > super_concept.objects.length then return false
+		return super_concept.objects.has_all(objects)
+	end
+
+	redef fun to_s do return "{attributes}\n{objects}"
+end
+
+# Concept Lattice
+#
+# Formal concepts are partially ordered with regard to inclusion of their extents
+# or (which is equivalent) inverse inclusion of their intent.
+class ConceptLattice[O: Object, A: Object]
+	super POSet[FormalConcept[O, A]]
+
+	# Build `self` from a set of formal `concepts`.
+	init from_concepts(concepts: Set[FormalConcept[O, A]]) do
+		for c in concepts do
+			add_node c
+		end
+		for c1 in concepts do
+			for c2 in concepts do
+				if c1 == c2 then continue
+				if not is_lower_neighbour(c1, c2, concepts) then continue
+				add_edge(c2, c1)
+			end
+		end
+	end
+
+	# Is `sub` the greatest lower bound of `sup` considering all `concepts`?
+	fun is_lower_neighbour(sub, sup: FormalConcept[O, A], concepts: Set[FormalConcept[O, A]]): Bool
+	do
+		if sub == sup then return false
+		if not sub.is_subconcept(sup) then return false
+		for concept in concepts do
+			if sub == concept then continue
+			if sup == concept then continue
+			if not sub.is_subconcept(concept) then continue
+			if not concept.is_subconcept(sup) then continue
+			return false
+		end
+		return true
+	end
+end
+
+var fc = new FormalContext[Int, String]
+fc.set_object_attributes(1, ["odd", "square"])
+fc.set_object_attributes(2, ["even", "prime"])
+fc.set_object_attributes(3, ["odd", "prime"])
+fc.set_object_attributes(4, ["even", "composite", "square"])
+fc.set_object_attributes(5, ["odd", "prime"])
+fc.set_object_attributes(6, ["even", "composite"])
+fc.set_object_attributes(7, ["odd", "prime"])
+fc.set_object_attributes(8, ["even", "composite"])
+fc.set_object_attributes(9, ["odd", "square", "composite"])
+fc.set_object_attributes(10, ["even", "composite"])
+
+var concepts = fc.formal_concepts
+for concept in concepts do
+	print concept
+end
+
+var cl = new ConceptLattice[Int, String].from_concepts(concepts)
+cl.show_dot
```

Comment:
it is often not a good idea to have a `main` in a library since it may cause issues with the real main of the client-program if there is linearization or call to super